Add tutorial for adding users to ILS db (#24)
inveniosoftware/docs-invenio-ils: docs/customize/users.md

> # Adding Users to Invenio ILS instance

Reviewer: We branded the products InvenioILS and InvenioRDM without spaces:
```suggestion
# Adding users to InvenioILS instance
```
Can you change it everywhere?

> This guide provides step-by-step instructions for adding users to a custom Invenio App ILS instance using:

Reviewer: I am not 100% sure that readers know what is `Invenio App ILS`, given that we refer to it as `InvenioILS`:
```suggestion
This guide provides step-by-step instructions for adding users to your own InvenioILS instance using two possible different methods:
```

> - An existing database
> - LDAP synchronization

Reviewer: You are presenting the tutorial in the opposite order (first LDAP, then DB):

```suggestion
1. LDAP synchronization
2. An existing external database
```

!!! Note
    The following guides are just an example to give you an idea of how you can get started on existing users to your Invenio ILS instance!

    Similar scripts can also be used to port other cataloging data like Documents, E-Items, Purchase Orders and more!

This process is suitable for administrators who have just set up a new instance and want to onboard all users to the ILS database.

### Prerequisites

> - A running instance of Invenio App ILS.

Reviewer: Global change:
```suggestion
- A running instance of InvenioILS.
```

- Administrative access to the Invenio App ILS instance.

> - Python installed on your system.

Reviewer: If I have the first, the second prerequisite must be there... I would remove the Python part

> ## Adding Users to Invenio App ILS Instance using LDAP Sync

Reviewer: ```suggestion
## 1. Adding users using LDAP sync
```

    for entry in conn.entries:
        # Extract user details
        user_data = {field: getattr(entry, ldap_field, [None])[0] for field, ldap_field in LDAP_USER_FIELDS_MAPPING.items()}

        # Check if the user already exists
        user = User.query.filter_by(email=user_data['email']).first()

        if user is None:
            # Create a new user
            user = User(
                email=user_data['email'],
                username=user_data['username'],
                active=True,
                profile=dict(
                    full_name=f"{user_data['first_name']} {user_data['last_name']}",
                ),
            )
            db.session.add(user)
            print(f"Added new user: {user.email}")
        else:
            # Update existing user
            user.username = user_data['username']
            user.profile['full_name'] = f"{user_data['first_name']} {user_data['last_name']}"
            print(f"Updated existing user: {user.email}")

    # Commit the changes
    db.session.commit()
    print("LDAP synchronization complete.")

if __name__ == '__main__':
    ldap_sync()
```

> ## Adding Users to Invenio App ILS Instance from a Custom Database

Reviewer: ```suggestion
## Adding users from an existing external database
```

### Step 1: Install Required Packages

Ensure that the required Python packages are installed. You can install them using pip:

```bash
pip install sqlalchemy
```

> Make sure you have access to the custom user database.

Reviewer: ```suggestion
Make sure you have access to the external user database.
```

### Step 2: Configure Database Connection

Create a configuration file or environment variables to store your database connection details. Here's an example of setting up a configuration file:

````python
CUSTOM_DB_URI = 'postgresql://user:password@localhost/customdb'
INVENIO_DB_URI = 'postgresql://invenio:password@localhost/invenio'```
````

> ### Step 3: Create a Script to Port Users

Reviewer: ```suggestion
### Step 3: Create a script to import users
```

> Create a Python script to read users from the custom database and add them to the Invenio App ILS database. Save the following script as **port_users.py**:

Reviewer: ```suggestion
Create a Python script to read users from the external database and add them to the InvenioILS database. Save the following script as **import_users.py**:
```

```python
import os
from sqlalchemy import create_engine, MetaData, Table
from sqlalchemy.orm import sessionmaker
from invenio_accounts.models import User
from invenio_db import db
from invenio_app.factory import create_api

# Load configuration

> custom_db_uri = os.getenv('CUSTOM_DB_URI', 'postgresql://user:password@localhost/customdb')

Reviewer: Apply everywhere
```suggestion
external_db_uri = os.getenv('CUSTOM_DB_URI', 'postgresql://user:password@localhost/customdb')
```

invenio_db_uri = os.getenv('INVENIO_DB_URI', 'postgresql://invenio:password@localhost/invenio')

# Create SQLAlchemy engines and sessions
custom_engine = create_engine(custom_db_uri)
invenio_engine = create_engine(invenio_db_uri)

CustomSession = sessionmaker(bind=custom_engine)
InvenioSession = sessionmaker(bind=invenio_engine)

custom_session = CustomSession()
invenio_session = InvenioSession()

# Define the custom user table (adjust to match your schema)
metadata = MetaData()
users_table = Table('users', metadata, autoload_with=custom_engine)

> def port_users():

Reviewer: ```suggestion
def import_users():
```

    for custom_user in custom_users:
        # Extract user details (adjust to match your schema)
        email = custom_user.email
        username = custom_user.username
        first_name = custom_user.first_name
        last_name = custom_user.last_name

        # Check if the user already exists in Invenio ILS DB
        user = invenio_session.query(User).filter_by(email=email).first()

        if user is None:
            # Create a new user
            user = User(
                email=email,
                username=username,
                active=True,
                profile=dict(
                    full_name=f"{first_name} {last_name}",
                ),
            )
            invenio_session.add(user)
            print(f"Added new user: {email}")
        else:
            # Update existing user
            user.username = username
            user.profile['full_name'] = f"{first_name} {last_name}"
            print(f"Updated existing user: {email}")

    # Commit the changes
    invenio_session.commit()
    print("User porting complete.")

if __name__ == '__main__':
    # Initialize Invenio app context
    app = create_api()
    with app.app_context():

>         port_users()

Reviewer: ```suggestion
        import_users()
```

    # Close sessions
    custom_session.close()
    invenio_session.close()
```

### Additional Tips

> You may schedule the script to run periodically (e.g., using a cron job) if you need to keep the Invenio user database in sync with your custom database/LDAP directory.

Reviewer: ```suggestion
You may schedule the script to run periodically (e.g., using a cron job) if you need to keep the InvenioILS user database in sync with your external database or LDAP directory.
```
